validator/c/src/lib.rs

```diff
@@ -490,53 +530,97 @@ fn export(
-    // substrait_validator_free_exported(), so we don't have to rely on the
-    // user to save this data.
+    // Save the length and capacity of the vector to the start of said
+    // vector, so we can recover them later.
     let len: u64 = data.len().try_into().unwrap();
     data[..8].clone_from_slice(&len.to_ne_bytes());
-
-    // Also pass the length to the user, if they wanted to know about it.
-    if let Some(size) = unsafe { size.as_mut() } {
-        // Note that the true size is smaller than the vector because of the
-        // 8-byte length prefix and the 1-byte null-termination character.
-        *size = len - 9;
-    }
+    let capacity: u64 = data.capacity().try_into().unwrap();
+    data[8..16].clone_from_slice(&capacity.to_ne_bytes());
 
     // Get the pointer to the vector, and relinquish ownership.
     let ptr = data.as_mut_ptr();
     std::mem::forget(data);
 
-    // Advance the pointer beyond the bytes that we're using to store the size
-    // of the vector.
-    unsafe { ptr.add(8) }
+    // UNSAFE: advance the pointer beyond the bytes that we're using to store
+    // the size of the vector. This assumes that advancing by 16 bytes doesn't
+    // advance beyond the end of the buffer, which should not be possible, as
+    // the buffer is at least 17 bytes long (8 bytes length, 8 bytes capacity,
+    // and null termination byte).
+    unsafe { ptr.add(16) }
+}
+
+/// Frees memory associated with an exported buffer. No-op if given a nullptr.
+#[no_mangle]
+pub extern "C" fn substrait_validator_free_exported(data: *mut u8) {
+    // Don't do anything if the user passed nullptr.
+    if data.is_null() {
+        return;
+    }
+
+    // UNSAFE: recover the pointer to the vector data. Assumes that the pointer
+    // was (ultimately) created using export(), in which case this just
+    // reverses the pointer arithmetic done at the end of its body.
+    let buffer_ptr = unsafe { data.sub(16) };
+
+    // UNSAFE: recover the vector length from the first 8 bytes. Assumes that
+    // these 8 bytes are readable.
+    let length_ptr = buffer_ptr;
+    let length = u64::from_ne_bytes(
+        unsafe { std::slice::from_raw_parts(length_ptr, 8) }
+            .try_into()
+            .unwrap(),
+    );
+    let length = usize::try_from(length).unwrap();
+
+    // UNSAFE: recover the vector capacity from the next 8 bytes. Assumes that
+    // these 8 bytes are readable.
+    let capacity_ptr = unsafe { buffer_ptr.add(8) };
+    let capacity = u64::from_ne_bytes(
+        unsafe { std::slice::from_raw_parts(capacity_ptr, 8) }
+            .try_into()
+            .unwrap(),
+    );
+    let capacity = usize::try_from(capacity).unwrap();
+
+    // UNSAFE: recover the vector and drop it. Assumes that the recovered
+    // pointer, length, and capacity do indeed form the raw parts of a valid
+    // Vec.
+    let vec = unsafe { Vec::from_raw_parts(buffer_ptr, length, capacity) };
+    drop(vec);
```

Comment:
You don't need this `drop` but I like it.

Reply:
Haha, yeah, more explicit more better, right? It was there before too, I just pulled it out of the `unsafe {}`.